File: sdks/js/src/magic.ts
/**
 * Magic link client — wraps the /api/v1/magic/send and
 * /api/v1/magic/verify endpoints.
 *
 * All camelCase input options are converted to snake_case before being sent
 * to the API. All snake_case keys in the API response are converted to
 * camelCase before being returned to the caller.
 */

import { MailGuardClient } from './client';
import type {
  MagicLinkSendOptions,
  MagicLinkSendResult,
  MagicLinkVerifyResult,
} from './types';

/** Raw API response shape for GET /api/v1/magic/verify/{token} (snake_case). */
interface RawMagicLinkVerifyResult {
  valid: boolean;
  email_hash: string;
  project_id: string;
  purpose: string;
  redirect_url: string;
}

export class MagicLinkClient extends MailGuardClient {
  /**
   * Send a magic link to the given email address.
   *
   * @param options - email, purpose, and redirectUrl (camelCase) — redirectUrl
   *   is automatically converted to redirect_url (snake_case) in the request.
   * @returns MagicLinkSendResult with status: 'sent'
   * @throws RateLimitError  on HTTP 429
   * @throws SandboxError    on HTTP 403 with sandbox_key_in_production
   * @throws InvalidKeyError on HTTP 401
   * @throws MailGuardError  for all other errors
   */
  async send(options: MagicLinkSendOptions): Promise<MagicLinkSendResult> {
    return this.request<MagicLinkSendResult>('POST', '/api/v1/magic/send', {
      email: options.email,
      purpose: options.purpose,
      redirect_url: options.redirectUrl,
    });
  }

  /**
   * Verify a magic link token.
   *
   * @param token - Raw magic link token extracted from the URL
   * @returns MagicLinkVerifyResult with camelCase keys
   * @throws ExpiredError    on HTTP 410 (token expired or already used)
   * @throws InvalidKeyError on HTTP 401
   * @throws MailGuardError  for all other errors
   */
  async verify(token: string): Promise<MagicLinkVerifyResult> {
    const raw = await this.request<RawMagicLinkVerifyResult>(
      'GET',
      `/api/v1/magic/verify/${encodeURIComponent(token)}`,
    );

    return {
      valid: raw.valid,
      emailHash: raw.email_hash,
      projectId: raw.project_id,
      purpose: raw.purpose,
      redirectUrl: raw.redirect_url,
    };
  }
}

File: sdks/python/mailguard/magic.py
"""
Magic link client — wraps the /api/v1/magic/send and
/api/v1/magic/verify endpoints.

All API response fields are kept in snake_case, matching Python conventions.
Returns TypedDicts for IDE autocompletion without external type stubs.
"""

from .client import MailGuardClient
from .async_client import AsyncMailGuardClient
from .types import (
    MagicLinkSendOptions,
    MagicLinkSendResult,
    MagicLinkVerifyResult,
)


class MagicLinkClient(MailGuardClient):
    """Synchronous magic link client."""

    def send(self, options: MagicLinkSendOptions) -> MagicLinkSendResult:
        """
        Send a magic link to the given email address.

        :param options: MagicLinkSendOptions with ``email``, ``purpose``,
                        and ``redirect_url``.
        :returns: MagicLinkSendResult TypedDict with ``status: 'sent'``.
        :raises RateLimitError:  on HTTP 429
        :raises SandboxError:    on HTTP 403 with sandbox_key_in_production
        :raises InvalidKeyError: on HTTP 401
        :raises MailGuardError:  for all other errors
        """
        raw = self._request(
            "POST",
            "/api/v1/magic/send",
            {
                "email": options["email"],
                "purpose": options["purpose"],
                "redirect_url": options["redirect_url"],
            },
        )
        return MagicLinkSendResult(status=raw["status"])

    def verify(self, token: str) -> MagicLinkVerifyResult:
        """
        Verify a magic link token.

        :param token: Raw magic link token extracted from the URL.
        :returns: MagicLinkVerifyResult TypedDict with ``valid``,
                  ``email_hash``, ``project_id``, ``purpose``,
                  and ``redirect_url``.
        :raises ExpiredError:    on HTTP 410 (token expired or already used)
        :raises InvalidKeyError: on HTTP 401
        :raises MailGuardError:  for all other errors
        """
        import urllib.parse

        raw = self._request(
            "GET",
            f"/api/v1/magic/verify/{urllib.parse.quote(token, safe='')}",
        )
        return MagicLinkVerifyResult(
            valid=raw["valid"],
            email_hash=raw["email_hash"],
            project_id=raw["project_id"],
            purpose=raw["purpose"],
            redirect_url=raw["redirect_url"],
        )


class AsyncMagicLinkClient(AsyncMailGuardClient):
    """Asynchronous magic link client."""

    async def send(self, options: MagicLinkSendOptions) -> MagicLinkSendResult:
        """
        Async version of MagicLinkClient.send().

        :param options: MagicLinkSendOptions with ``email``, ``purpose``,
                        and ``redirect_url``.
        :returns: MagicLinkSendResult TypedDict.
        """
        raw = await self._request(
            "POST",
            "/api/v1/magic/send",
            {
                "email": options["email"],
                "purpose": options["purpose"],
                "redirect_url": options["redirect_url"],
            },
        )
        return MagicLinkSendResult(status=raw["status"])

    async def verify(self, token: str) -> MagicLinkVerifyResult:
        """
        Async version of MagicLinkClient.verify().

        :param token: Raw magic link token extracted from the URL.
        :returns: MagicLinkVerifyResult TypedDict.
        """
        import urllib.parse

        raw = await self._request(
            "GET",
            f"/api/v1/magic/verify/{urllib.parse.quote(token, safe='')}",
        )
        return MagicLinkVerifyResult(
            valid=raw["valid"],
            email_hash=raw["email_hash"],
            project_id=raw["project_id"],
            purpose=raw["purpose"],
            redirect_url=raw["redirect_url"],
        )
